Bump to ospray 3.0.0

demarle opened the pull request:
### Description

<!-- Please include a summary of the change and any other information and instructions to help reviewers understand the work -->

This change bumps the develop branch of visit from OSPRay 2 to OSPRay 2

<!-- Note that all the checklist items in various bulleted lists below end with ~~ characters.
     This is to facilitate striking them out when they do not apply.
     One just has to add ~~ characters just before the opening square bracket [ -->

### Type of change

<!-- Please check one of the boxes below -->

* [ ] Bug fix~~
* [ ] New feature~~
* [ ] Documentation update~~
* [x ] Other~~ <!-- please explain with a note below -->

Dependency version update.

### How Has This Been Tested?

<!-- Please describe the tests you've added or any tests that already cover this change. Include relevant information, such as which operating system you tested on. -->

Informal testing on my local workstation and on Aurora. 
Local testing was load a pvti, verify that it looks about the same with and without ospray enabled for both a contour plot and volume plot.

### Reminders:

- Please follow the [style guidelines][1] of this project.
- Please perform a self-review of your code before submitting a PR and asking others to review it.
- Please assign reviewers (see [VisIt's PR procedures][2] for more information).

### Checklist:

<!-- For items in this checklist that do not apply, simply insert two tilde chars, `~~`, just ahead of the left bracket char, `[` at the beginning of a line. Each line ends with two tilde chars to make doing such ~~strikeouts~~ easy. -->

- [ ] I have commented my code where applicable.~~
- [ ] I have updated the release notes.~~
- [ ] I have made corresponding changes to the documentation.~~
- [ ] I have added debugging support to my changes.~~
- [ ] I have added tests that prove my fix is effective or that my feature works.~~
- [ ] I have confirmed new and existing unit tests pass locally with my changes.~~
- [ ] I have added new baselines for any new tests to the repo.~~
- [ ] I have NOT made any changes to [*protocol* or *public interfaces*][3] in an RC branch.~~

[1]: https://visit-sphinx-github-user-manual.readthedocs.io/en/develop/dev_manual/StyleGuide.html
[2]: https://visit-sphinx-github-user-manual.readthedocs.io/en/develop/dev_manual/pr_create.html#reviewers
[3]: https://visit-sphinx-github-user-manual.readthedocs.io/en/develop/dev_manual/RCDevelopment.html#communication-protocols-and-public-apis

demarle committed:
patch VTK 9.1.0 for OSPRay 2

new patch is
https://gitlab.kitware.com/vtk/vtk/-/merge_requests/10529
backported to VTK 9.1.0

demarle committed:
build script change for ospray 3.0.0

biagas commented on src/CMake/FindOSPRay.cmake:
```diff
@@ -175,23 +175,29 @@ else() # ospray > 1.6.1
-                       openvkl_module_cpu_device_4
-                       rkcommon
-                       tbbmalloc)
+                       libembree4.so.4
```
Should not be listing .so.V here, those won't work for native Windows builds where there are .lib and .dll extensions.

biagas replied:
BTW, I am attempting builds on Linux and Windows and will report back what I discover about this particular change.

demarle replied:
Thanks! Yeah for whatever reason my local setup seemed to need the full lib name. I'll take them out and remove the comment about it on on line 179.

biagas replied:
Our SET_UP_THIRD_PARTY function is performing a `find_library` for all the listed `LIBS`.
Due to missing plain old '.so' versions of some libs in the OSPRay install, the `find_library` call is failing. No wonder you need to list the whole file name. Unfortunately if we use the full file name, would need a different list for Windows, possibly one for MacOS too.

I suggest a change to use simple `find_package` for 3.x versions of OSPRay.
Something like this:
```
    if(EXISTS ${VISIT_OSPRAY_DIR}/lib64)
        set(LIB lib64)
    else()
        set(LIB lib)
    endif()

    set(ospray_DIR ${VISIT_OSPRAY_DIR}/${LIB}/cmake/ospray-${OSPRAY_VERSION})
    find_package(ospray NO_MODULE PATHS ${ospray_DIR})
```

Would need to add logic for installing the ospray libs alongside visit, and possibly changes to a few visit targets, haven't got that far yet.

biagas replied:
@demarle Could you try this version of FindOSPRay.cmake and let me know if you run into problems?

[FindOSPRay.zip](https://github.com/visit-dav/visit/files/13171624/FindOSPRay.zip)

It is only installing ospray and rkcommon libs alongside VisIt, as that appears to be all that was needed (on our linux system anyways)

biagas replied:
My simplified find logic doesn't work on Windows, so still investigating.

biagas replied:
@demarle may I ask how you came up with the list of ospray libraries that were needed? It surprised me when I modified the find module to only find ospray and rkcommon that VisIt appeared to work correctly on one of our linux machines. But perhaps some system libs were found and used at runtime.

demarle replied:
Sure thing!

When I built it locally I scanned the OSPRay install directory and took out the libraries that I knew were not needed by apps that use ospray. Ie omitting stuff related to ospray's own examples and tests. I then just ran visit surface and volume rendering with ospray to check if I missed something important. I might have left something extra in but I'm really surpised by your finding that only rkcommon and ospray were needed.

OSPRay always uses embree and volume rendering adds the dependency on openVKL. ispcrt (ISPC Run Time) is a new dependency as of ospray 2.12 when the infrastructure for GPU started to land. sycl is a new dependency for the GPU features in 3.0.

OSPRay tends to use runtime loading (dlopen) so I suppose it could be the case that it won't complain if openVKL is missing until volume rendering is used but I would be surprised. Perhaps it was the case that bits were statically linked in? Tomorrow I'll investigate some more.

biagas replied:
I went ahead and modified FindOSPRay to use file(GLOB ...) to find all the libraries in the ospray/lib(64) dir for installation alongside VisIt.  More of the libraries were definitely needed on Windows.
```
   find_package(ospray ${OSPRAY_VERSION} REQUIRED
                 PATHS ${VISIT_OSPRAY_DIR}
                 PATH_SUFFIXES lib/cmake/opsray-${OSPRAY_VERSION}
                               lib64/cmake/ospray-${OSPRAY_VERSION}
                 NO_MODULE
                 NO_DEFAULT_PATH)

    if(ospray_FOUND)
        set(HAVE_LIBOSPRAY true)
        add_definitions(-DHAVE_OSPRAY)
        if(VISIT_INSTALL_THIRD_PARTY)
            # since all the libs needed for VisIt at runtime aren't
            # enumerated in the ospray targets from find_package,
            # just install all of them 
            set(LIB_SEARCH_PATH ${VISIT_OSPRAY_DIR}/lib)
            if(EXISTS ${VISIT_OSPRAY_DIR}/lib64)
                set(LIB_SEARCH_PATH ${VISIT_OSPRAY_DIR}/lib64)
            endif()
            if(NOT ospray_lib_libs)
                file(GLOB ospray_lib_libs
                     LIST_DIRECTORIES FALSE
                     ${LIB_SEARCH_PATH}/*)
            endif()
            unset(LIB_SEARCH_PATH)
            #  install libraries
            foreach(lib ${ospray_lib_libs})
                THIRD_PARTY_INSTALL_LIBRARY(${lib})
            endforeach()

            if(NOT VISIT_HEADERS_SKIP_INSTALL)
                THIRD_PARTY_INSTALL_INCLUDE(ospray ${OSPRAY_INCLUDE_DIR})
            endif()
        endif()
    endif()

```

What do you think? 

Works so far on our Linux and Windows as far as build and install.

Ospray rendering seems to be working fine on Linux, but not on Windows. Still investigating on Windows.

demarle replied:
I like it. Two suggestions:
1) Perhaps we could add an "exclude list" at some point for any libraries we know not to be useful to keep minimize size.
2) When OSPRay's superbuild's cmake is told to download rather than build its own dependencies it can end up with a mixture of "lib/" and "lib64/" files. If we run into that inside VisIt's build of OSPRay we might want to revisit the either or we've got above and just append to LIB_SEARCH_PATH.

biagas replied:
Could you make that change to FindOSPRay then on your branch?

demarle replied:
Yep will do.

demarle replied:
Done: ie, I swapped in your globbed lib find and added the line that removes the two libs which VisIt doesn't need. 

Another, probably better, approach would be to modify bv_ospray.sh line 345 or so to tell ospray 3's superbuild to turn off apps and other parts of OSPRay that VisIt doesn't need (like build_ospray_in_source still does for ospray 1.x). To the best of my knowledge visit doesn't use openimagedenoise yet so you could save some slightly more substantial compile time and install space that way.

I checked on the lib/lib64 thing and I think that is a red herring.

biagas replied:
Thanks for testing that out, and removing the libs we don't need.
I like the idea of modifying bv_ospray to prevent unnecessary libs/executables from being built in the first place.
I can enter a ticket to take care of this at a later time.
Are you ready for this to merge (it works on Linux)? Or do you need the Windows version to work first?

biagas commented on src/tools/dev/scripts/bv_support/bv_visit.sh:
```diff
@@ -345,6 +345,15 @@ function build_visit
         FEATURES="${FEATURES} -DCPACK_RPM_SPEC_MORE_DEFINE:STRING=\"%global_python_bytecompile_errors_terminate_build 0\""
     fi
 
+    FEATURES="${FEATURES} -DENABLE_EXAMPLES:BOOL=OFF"
```
Not sure we want to turn these off in this script. @brugger1, @cyrush thoughts?

brugger1 replied:
We should not turn these off. This change should be backed out.

demarle replied:
Oops, sorry about that those were to facilitate a local build and I neglected to take them out before the MR. I'll remove them and force push.

demarle replied:
done

biagas commented:
I built ospray 3 on Windows, built VTK 9.2.6 against ospray 3, then built VisIt with them.
Any type of ospray rendering yields a blank window.
Do you think its related to: https://github.com/ospray/ospray/issues/573?

demarle commented:
https://github.com/ospray/ospray/issues/573 could be related, the symptoms sure are similar. I'll hop back on that VTK issue tomorrow. When I poked at it last time the OSPRay build itself was working (it's tests and examples were fine), but Kitware's ParaView build also produced blank images like yours. My two guesses are something related to blending of the ospray framebuffer onto the VTK Renderer/layer or something wrong related to runtime library linking.

demarle committed:
get all of OSPRay's libs reliably

thanks Kathleen B for the GLOB approach

changed build_visit's build of OSPRay's superbuild to skip
 OSPRay tests and apps that VisIt won't need and
 OpenImageDenoise that VisIt does not yet expose

biagas commented:
On Windows, I played around with different versions of VTK and ospray, compiling with MSVC2022. 
I grabbed patches for VTK from linux versions built with the changes to bv_vtk.
Here's what I discovered:

**VTK-9.1, ospray 2.8** general ospray rendering works, volumue plot with ospray works

**VTK-9.2.6, ospray 2.8** General ospray rendering does not work, Volume plot with OSPRay works

**VTK-9.2.6, ospray 3.0**  Niether ospray methods work

demarle commented:
Thanks Kathleen. 

I'll take a first stab now at that bv_ospray.sh script change.
I'll continue debugging the windows ospray > problem tonight. 

In general it "works for me" on Linux and builds at least on Windows so I'm think I'm OK with it getting merged into the devel branch if you don't mind.

demarle commented:
Pushed the revision just now. It works for me (on Linux). I'll fix the OSPRay on windows problem, in all probability at the VTK level, and push a new MR here if necessary.

biagas committed:
Merge branch 'develop' into bump-to-ospray-3.0.0

biagas commented:
Will go ahead with the merge so this can go in the 3.4 release. Can work on the Windows fixes later.

demarle commented:
A likely fix for the Windows issue is under review over in VTK here:
https://gitlab.kitware.com/vtk/vtk/-/merge_requests/10688

biagas commented:
> A likely fix for the Windows issue is under review over in VTK here: https://gitlab.kitware.com/vtk/vtk/-/merge_requests/10688

That's great. I might be able to patch my build on Windows with that fix before the release.
